### ISSUE TITLE
Sparse arrays do not support all of Scipy's slicing options

### ISSUE THREAD

**Issue by cjnolet** [cupy/cupy#3178]
I encountered this behavior in the most recent (CuPy 8.0.0 at the time of writing this) branch when attempting to call `dask.array.from_array()` with a `csr_matrix`. It appears that Dask is attempting to do some slicing that is supported in Scipy's `csr_matrix` but is not supported by CuPy. 

The following is a reproducible example:

```
>>> import cupy as cp
>>> import dask.array
>>> 
>>> a = cp.sparse.random(100, 100, format='csr')
>>> b = dask.array.from_array(a)
>>> b.compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/share/workspace/dask/dask/base.py", line 165, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/share/workspace/dask/dask/base.py", line 436, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/share/workspace/dask/dask/threaded.py", line 81, in get
    **kwargs
  File "/share/workspace/dask/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/share/workspace/dask/dask/local.py", line 316, in reraise
    raise exc
  File "/share/workspace/dask/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/share/workspace/dask/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/share/workspace/dask/dask/array/core.py", line 104, in getter
    c = a[b]
  File "/share/workspace/cupy/cupyx/scipy/sparse/compressed.py", line 509, in __getitem__
    raise ValueError('unsupported indexing')
ValueError: unsupported indexing
```

Here's the same code using Scipy:

```
>>> import scipy
>>> 
>>> a = scipy.sparse.random(100, 100, format='csr')
>>> b = dask.array.from_array(a)
>>> b.compute()
array(<100x100 sparse matrix of type '<class 'numpy.float64'>'
	with 100 stored elements in Compressed Sparse Row format>, dtype=object)
```

When using a CuPy `coo_matrix`:

```
>>> a = cp.sparse.random(100, 100, format='coo')
>>> b = dask.array.from_array(a)
>>> b.compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/share/workspace/dask/dask/base.py", line 165, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/share/workspace/dask/dask/base.py", line 436, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/share/workspace/dask/dask/threaded.py", line 81, in get
    **kwargs
  File "/share/workspace/dask/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/share/workspace/dask/dask/local.py", line 316, in reraise
    raise exc
  File "/share/workspace/dask/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/share/workspace/dask/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/share/workspace/dask/dask/array/core.py", line 104, in getter
    c = a[b]
TypeError: 'coo_matrix' object is not subscriptable
```

As expected, the CuPy `csc_matrix` yields the same exception as CSR:

```
>>> a = cp.sparse.random(100, 100, format='csc')
>>> b = dask.array.from_array(a)
>>> b.compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/share/workspace/dask/dask/base.py", line 165, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/share/workspace/dask/dask/base.py", line 436, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/share/workspace/dask/dask/threaded.py", line 81, in get
    **kwargs
  File "/share/workspace/dask/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/share/workspace/dask/dask/local.py", line 316, in reraise
    raise exc
  File "/share/workspace/dask/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/share/workspace/dask/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/share/workspace/dask/dask/array/core.py", line 104, in getter
    c = a[b]
  File "/share/workspace/cupy/cupyx/scipy/sparse/compressed.py", line 509, in __getitem__
    raise ValueError('unsupported indexing')
ValueError: unsupported indexing
```

**Comment by jakirkham** [comment]
Just to simplify this a bit, the issue can be reproduced using CuPy 7.3.0 alone.

```python
In [1]: import cupy as cp                                                       

In [2]: a = cp.sparse.random(100, 100, format='csr')                            

In [3]: a[0:100, 0:100]                                                         
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-a8ed163789e5> in <module>
----> 1 a[0:100, 0:100]

/datasets/jkirkham/miniconda/envs/rapids13dev/lib/python3.7/site-packages/cupyx/scipy/sparse/compressed.py in __getitem__(self, slices)
    247                 return self._get_major_slice(major)
    248 
--> 249         raise ValueError('unsupported indexing')
    250 
    251     def _get_single(self, major, minor):

ValueError: unsupported indexing
```

**Comment by jakirkham** [comment]
Should add indexing on the first axis works. Indexing on the second axis is problematic.

```python
In [1]: import cupy as cp                                                       

In [2]: a = cp.sparse.random(100, 100, format='csr')                            

In [3]: a[0:100]                                                                
Out[3]: <cupyx.scipy.sparse.csr.csr_matrix at 0x7fc7fb09cd10>

In [4]: a[:, 0:100]                                                             
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-fb183ec4eed3> in <module>
----> 1 a[:, 0:100]

/datasets/jkirkham/miniconda/envs/rapids13dev/lib/python3.7/site-packages/cupyx/scipy/sparse/compressed.py in __getitem__(self, slices)
    247                 return self._get_major_slice(major)
    248 
--> 249         raise ValueError('unsupported indexing')
    250 
    251     def _get_single(self, major, minor):

ValueError: unsupported indexing
```

**Comment by pentschev** [comment]
@asi1024 as per our discussion during the PFN-NVIDIA meeting, would you be able to tell us here if there's any reason for this feature not to be allowed in CuPy? It seems that this is just something that nobody has had a chance to work on so far, but it would be great to have some confirmation.

**Comment by crumleyc** [comment]
I am currently a grad student working on my thesis. I am running into this same issue where the error is saying that the CuPy sparse array isn't subscript-able. I am wanting to show interest in working on this if it is appropriate. I haven't worked on any open source project, especially something like this. If you think this is something I could help out with, please point me in the right direction, if not that is fine too.

**Comment by asi1024** [comment]
@pentschev There is no reason for this indexing not to be supported.

**Comment by asi1024** [comment]
Current implementation raises a `ValueError`, but it should raise `NotImplementedError`. Sorry for being confusing.

**Comment by cjnolet** [comment]
@crumleyc, we are needing this feature in Dask and RAPIDS currently. I think it would be great if you worked on this issue! Ideally, this change will be included in the 8.0.0 release.

As a first step, it will be helpful to familiarize yourself with Cupy’s [contribution guide](https://docs-cupy.chainer.org/en/stable/contribution.html), if you have not done so already. 

I tend to create a [pull request](https://docs-cupy.chainer.org/en/stable/contribution.html) early, so that I can solicit feedback as I work through changes. An additional benefit to this is that I am able to reference my changes directly when asking for help.

**Comment by cjnolet** [comment]
I have verified that the current implementation of #3486 fixes this issue.
